core/handler/amqp.go

```diff
@@ -66,29 +70,50 @@ func (h *handler) HandleAMQP(username, password, host, exchange, downlinkQueue s
 	if err != nil {
 		return err
 	}
+	go h.amqp()
+	return nil
+}
 
+func (h *handler) amqp() error {
 	ctx := h.Ctx.WithField("Protocol", "AMQP")
+	publisher := h.amqpClient.NewPublisher(h.amqpExchange)
+	err := publisher.Open()
+	if err != nil {
+		ctx.WithError(err).Error("Could not open publisher channel")
+		return err
+	}
+	defer publisher.Close()
+	go h.amqpUplink(ctx, publisher)
+	h.amqpEvt(ctx, publisher)
+	return nil
+}
 
-	go func() {
-		publisher := h.amqpClient.NewPublisher(h.amqpExchange)
-		err := publisher.Open()
+func (h *handler) amqpUplink(ctx ttnlog.Interface, publisher amqp.Publisher) {
+
+	for up := range h.amqpUp {
+		ctx.WithFields(ttnlog.Fields{
+			"DevID": up.DevID,
+			"AppID": up.AppID,
+		}).Debug("Publish Uplink")
+		err := publisher.PublishUplink(*up)
 		if err != nil {
-			ctx.WithError(err).Error("Could not open publisher channel")
-			return
+			ctx.WithError(err).Warn("Could not publish Uplink")
 		}
-		defer publisher.Close()
+	}
+}
 
-		for up := range h.amqpUp {
```

Comment:
I agree, though I copied it from the handler/mqtt.go where there is two goroutine so I though I will do the same. Changing on the next commit